src/azure-cli/azure/cli/command_modules/acs/custom.py

```diff
@@ -1475,8 +1482,30 @@ def _create_role_assignment(cmd, role, assignee,
 
     # If the cluster has service principal resolve the service principal client id to get the object id,
     # if not use MSI object id.
-    object_id = _resolve_object_id(
-        cmd.cli_ctx, assignee) if resolve_assignee else assignee
+    object_id = assignee
+    if resolve_assignee:
+        from azure.graphrbac.models import GraphErrorException
+        error_msg = "Failed to resolve service principal object ID: "
+        try:
+            object_id = _resolve_object_id(cmd.cli_ctx, assignee)
+        except GraphErrorException as ex:
+            if ex.response is not None:
+                error_code = getattr(ex.response, "status_code", None)
+                error_reason = getattr(ex.response, "reason", None)
+                internal_error = ""
+                if error_code:
+                    internal_error += str(error_code)
+                if error_reason:
+                    if internal_error:
+                        internal_error += " - "
+                    internal_error += str(error_reason)
+                if internal_error:
+                    error_msg += "({}) ".format(internal_error)
+            error_msg += ex.message
+            # this should be UserFault or ServiceError, but it is meaningless to distinguish them here
+            raise CLIError(error_msg)
+        except Exception as ex:  # pylint: disable=bare-except
+            raise CLIError(error_msg + str(ex))
```

Comment:
Could you use the specific error type instead of `CLIError`? For more details, please refer to this doc [error_handling_guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md)

Reply:
@FumingZhang helped contribute this part of code. I think the comment in L1505 explains why we don't distinguish the error category.

Reply:
>  this should be UserFault or ServiceError, but it is meaningless to distinguish them here

@FumingZhang Why is it unnecessary to distinguish between `UserFault` and `ServiceError` here?

Reply:
We are calling graph client here and this may return a 3xx or 4xx error (maybe 5xx also). We intend to display this error message in the form of warning, so we use `except CLIError` to catch this type of exception [where](https://github.com/Azure/azure-cli/pull/20477/files#diff-f1fc443c5eb2ddee110be5f06110930285573e1debf2640ab83dead88943a9a5R560) it calls this function.

I hope that the type of error can be better described to users, so in the previous suggestion I parsed the exception object raised by graph client and added the error code and reason to the error message. Here is a screenshot of an example of this. ![Screenshot 2021-11-24 140718](https://user-images.githubusercontent.com/81607949/143184457-f0419fc1-3b5f-4a7c-8f90-cb5ce0792a2d.png)

I respect that we should use more accurate error types to describe errors, but in this scenario, I suppose using error codes to distinguish error types will not provide users with more information.

Reply:
Got it, if this error is not raised out, but is only used to print logs, there is no need to subdivide the error type